Ean/device flow as default

ean opened the pull request:
@mpolden - please take a look at this

Commit:
Which issuer is used depends on the system you talk to

IsLoggedIn must map the current system to the issuer that system accept

bjorncs commented on client/go/auth0/auth0.go:
```diff
@@ -129,6 +129,7 @@ func (a *Auth0) IsLoggedIn() bool {
 	}
 
 	// Check if token is valid.
+	// TODO: Choose issuer based on system
```
This can be deduced Auth0 device flow config from api endpoint (https://api.vespa-external-cd.aws.oath.cloud:4443/auth0/v1/device-flow-config).

Commit:
Avoid leaking config settings from config_test/TestConfig to other tests

Commit:
Make Auth0 device flow Default authentication for deploy service

Fall back to using tenant api keys if no auth0 system is configured

This is to maintain backwards compatibility while users migrate over to Auth0
device flow. Falling back to tenant api keys issue a deprecation warning asking
the user to setup the Auth0 device flow, and will be removed at a later time.

mpolden commented on client/go/cmd/config.go:
```diff
@@ -179,19 +183,41 @@ func (c *Config) X509KeyPair(app vespa.ApplicationID) (KeyPair, error) {
+		if _, err := os.Stat(c.APIKeyPath(tenantName)); err != nil {
+			return false
+		}
+		fmt.Fprintln(os.Stderr, "Defaulting to tenant API key is deprecated. Use Auth0 device flow: 'vespa auth login' instead")
```
Should print to the variable `stderr`.

mpolden commented on client/go/cmd/config.go:
```diff
@@ -179,19 +183,41 @@ func (c *Config) X509KeyPair(app vespa.ApplicationID) (KeyPair, error) {
+
+	// If no Auth0 token is created, fall back to tenant api key, but warn that this functionality is deprecated
+	a, err := auth0.GetAuth0(c.AuthConfigPath(), getSystemName(), getApiURL())
+	if err != nil || !a.HasSystem() {
```
Why do we need all these checks? Shouldn't we simply use the API key if it exists and print a warning that the user should setup Auth0 with `vespa auth login`?

ean replied:
This was never meant as a permanent solution, more of a way to support the old behaviour for some releases to nudge people to the device flow. It will give people some time to migrate over, but I am happy to rip it all out if we think this functionality is not part of any automated systems somewhere.

mpolden commented on client/go/cmd/config.go:
```diff
@@ -179,19 +183,41 @@ func (c *Config) X509KeyPair(app vespa.ApplicationID) (KeyPair, error) {
+	// If no Auth0 token is created, fall back to tenant api key, but warn that this functionality is deprecated
+	a, err := auth0.GetAuth0(c.AuthConfigPath(), getSystemName(), getApiURL())
+	if err != nil || !a.HasSystem() {
+		if _, err := os.Stat(c.APIKeyPath(tenantName)); err != nil {
```
Consider using `util.PathExists`.

mpolden commented on client/go/cmd/helpers.go:
```diff
@@ -270,19 +250,11 @@ func getDeploymentOpts(cfg *Config, pkg vespa.ApplicationPackage, target vespa.T
-				return vespa.DeploymentOpts{}, errHint(err, "Deployment to cloud requires an API key. Try 'vespa api-key'")
-			}
+		if cfg.UseAPIKey(deployment.Application.Tenant) {
+			opts.APIKey, err = cfg.ReadAPIKey(deployment.Application.Tenant)
```
This error should be checked.
